src/arq/_saw.py

```diff
@@ -0,0 +1,133 @@
+# -*- coding: utf-8 -*-
+
+
+import gevent.queue
+import itertools
+import random
+import struct
+
+from binascii import hexlify
+from enum import IntEnum
+
+from ._utils import concurrently
+
+
+class PacketType(IntEnum):
+    data = 0
+    ackn = 1
+
+
+HEADER = struct.Struct('>BH')
+HEADER_SIZE = HEADER.size
+
+
+def sequence_numbers():
+    """Generate sequence numbers that wrap around."""
+    for i in itertools.count(start=1):
+        yield i % 0xffff
+
+
+def recvfrom(socket, peer, size):
+    """Receive a packet from a designated peer."""
+    data, p = socket.recvfrom(size)
+    while p != peer:
+        data, p = socket.recvfrom(size)
+    return data
+
+
+def stop_and_wait(socket, peer, iq, oq, prefix='', p=0.0):
+    """."""
+
+    aq = gevent.queue.Queue()
+
+    def recv():
```

Comment:
**TODO**: make this re-usable in servers, where we need to handle incoming packets from multiple peers.  This can be handled by using a queue per peer like so:

1.  `recv()` calls some caller-provided function for fetching the next packet;
2. client passes `socket.recvfrom` as the source for packets;
3. server passes `Queue.get` as the source for packets.

Reply:
Done (the client part).

Reply:
Done (all of it).  See `_io.py:udp_client` and `_io.py:udp_server`.